src/aspire/storage/starfile.py

```diff
@@ -1,173 +1,178 @@
+                    raise StarFileError(
+                        f"Attempted overwrite of existing data block: {gemmi_block.name}"
+                    )
+            elif block_has_loop:
+                if gemmi_block.name not in self.blocks:
+                    # initialize DF from list of lists
+                    self.blocks[gemmi_block.name] = pd.DataFrame(
+                        loop_data, columns=loop_tags, dtype=str
+                    )
+                else:
+                    # enforce unique keys
+                    raise StarFileError(
+                        f"Attempted overwrite of existing data block: {gemmi_block.name}"
+                    )
 
-            logger.info("StarFile parse complete")
+    def write(self, filepath):
+        """
+        Converts `blocks` to a gemmi Document and writes to a starfile at the given filepath.
+        """
+        # create an empty Document
+        _doc = cif.Document()
+        filepath = str(filepath)
+        for name, df in self.blocks.items():
+            # construct new empty block
+            _block = _doc.add_new_block(name)
+            # are we constructing a pair (df has 1 row) or a loop (df has >1 rows)
+            if len(df) == 0:
+                continue
+            if len(df) == 1:
+                for col in list(df):
+                    # key is the column label. value is value in the df at row 0, column col
+                    _block.set_pair(col, str(df.at[0, col]))
+            elif len(df) > 1:
+                # initialize loop with column names
+                _loop = _block.init_loop("", list(df.columns))
+                for row in df.values.tolist():
+                    row = [str(row[x]) for x in range(len(row))]
+                    _loop.add_row(row)
+
+        _doc.write_file(filepath)
+
+    def get_block_by_index(self, index):
+        """
+        Retrieve a DataFrame representing a star file block by its position in the starfile
+        """
+        return self.blocks[list(self.blocks.keys())[index]]
 
-        logger.info("Initializing StarFile object from data")
-        self.init_from_blocks(blocks)
-        logger.info(f"Created <{self}>")
+    def __getitem__(self, key):
+        """
+        Retrieve the star file block with name `key`
+        """
+        return self.blocks[key]
 
-    def init_from_blocks(self, blocks):
+    def __setitem__(self, key, value):
         """
-        Initialize a StarFile from a list of blocks
-        :param blocks: An iterable of StarFileBlock objects
-        :return: An initialized StarFile object
+        Pass in a Pandas Dataframe to add a block named `key` with values `value`
         """
-        for block in blocks:
-            self.blocks[block.name] = block
+        self.blocks[key] = value
 
     def __repr__(self):
-        return f"StarFile with {len(self.blocks)} blocks"
+        return "StarFile with blocks: " + ", ".join(list(self.blocks.keys()))
```

Comment:
do we need to call list here?

Reply:
yeah, i was confused too. `keys()` returns a `DictView` object I believe. had to cast to `list`

Reply:
nevermind. fixed